### ISSUE TITLE
Q: How can I apply/combine different ways of representation updates?

### ISSUE THREAD

**Issue by cycle20** [molstar/molstar#264]
I found various examples to change representations in the source tree of Mol*.
I do not want to list them here - however it would be worth some details in the documentations.
(I hope I can dedicate some time to it in the near future.)
It includes various ways of accessing structures, models, hierarchies and other internal stuffs of Mol* - it seems very advanced to me...
Furthermore I got from you a very specific implementation-hint to add membrane representation programmatically. Thank you for it again. https://github.com/molstar/molstar/issues/237#issue-957238906

I could work out a solution that can colorize a specific series of residues in each chain of a loaded model. This approach worked - some days ago :)
Then I started to refactor my code and integrate the membrane representation and the coloring stuff into a single “workflow”. Finally it became a nightmare and I was completely lost in my different versions and branches.

### Basic blocks of my approach:

1. Download of data
2. Parse of trajectory
3. Apply of preset of membrane orientation
   _// steps 1-3 based on your suggestion:_ https://github.com/molstar/molstar/issues/237#issuecomment-890350752
4. Then I access the created structure to re-use it for creation of polymer, ligand, water components and to build their representations.
   _this the result of my try-and-error experiments and/or some sample from somewhere:_
```typescript
   const structure: StateObjectRef<PluginStateObject.Molecule.Structure> =
            this.plugin.managers.structure.hierarchy.current.models[0].structures[0].cell;
```
5. Creation of components in _step 4_.
6. Building representations of them.
    _probably from your proteopedia example:_ https://github.com/molstar/molstar/blob/master/src/examples/proteopedia-wrapper/index.ts#L359
7. Applying my update process of representation to paint selected residue series in chains

### My questions:
My suspicion is that I messed up the transformations somewhere.
Should I `commit` my updates only once or each representation `update` has to be performed by a new `StateBuilder` and requires a separate `commit`?
Do you have any debug suggestions/hints to discover similar _silent_ errors?
Can you review my code snippets and provide some advice?

### Code snippets:

```ts
type Region = {
    auth_ids: number[],
//    region_type: string,
    color: [ number, number, number ]
}
interface ChainDescriptor {
    chain_id: string,
    regions: Region[],
    seq: string,
    type: string
}

interface Descriptor {
    chains: ChainDescriptor[],
    'format': 'mmcif' | 'pdb'
}

async loadCustomRepresentation(url: string, descriptor: Descriptor) {
    const isBinary = false;

    // David Sehnal's version
    const data = await this.plugin.builders.data.download({ url: url, label: 'LABEL', isBinary: isBinary }, { state: { isGhost: true }});
    const trajectory = await this.plugin.builders.structure.parseTrajectory(data, descriptor.format);
    await this.plugin.builders.structure.hierarchy.applyPreset(
        trajectory, 'default', { representationPreset: 'preset-membrane-orientation' as any });

    //
    // Q: risky hack to access the structure?
    //
    const structure: StateObjectRef<PluginStateObject.Molecule.Structure> =
        this.plugin.managers.structure.hierarchy.current.models[0].structures[0].cell;
        
    const components = {
        polymer: await this.plugin.builders.structure.tryCreateComponentStatic(structure, 'polymer'),
        ligand: await this.plugin.builders.structure.tryCreateComponentStatic(structure, 'ligand'),
        water: await this.plugin.builders.structure.tryCreateComponentStatic(structure, 'water'),
    };

    const builder = this.plugin.builders.structure.representation;
    const update = this.plugin.build();
    if (components.polymer) builder.buildRepresentation(update, components.polymer, { type: 'cartoon', typeParams: { alpha: 0.5 } }, { tag: 'polymer' });
    if (components.ligand) builder.buildRepresentation(update, components.ligand, { type: 'ball-and-stick' }, { tag: 'ligand' });
    if (components.water) builder.buildRepresentation(update, components.water, { type: 'ball-and-stick', typeParams: { alpha: 0.6 } }, { tag: 'water' });
    
    //
    // Q: Do I need this commit as well...?
    //
    
    //await update.commit();

    for (let chain of descriptor.chains) {
        for (let region in chain.regions) {
            //
            // Q: ...or better to build transformations on a new(?) root?
            //
            //const update = this.plugin.build();

            this.applyColor(chain, region, update.to(structure));
        }
    }

    await update.commit();
    //
    // reset the camera because the membranes render 1st and the structure might not be fully visible
    //
    requestAnimationFrame(() => this.plugin.canvas3d?.requestCameraReset());
}

applyColor(chain: ChainDescriptor, region_key: any, update: StateBuilder.To<any, any>) {
    const region: Region = chain.regions[region_key];
    const color: Color = Color.fromArray(region.color, 0);
    const label: string = `${chain.chain_id} - ${region.color}`;

    const query: Expression = this.getQuery(chain.chain_id, region.auth_ids);

    // based on https://github.com/molstar/molstar/issues/209
    update
        .apply(StateTransforms.Model.StructureSelectionFromExpression, { label: label, expression: query })
        .apply(StateTransforms.Representation.StructureRepresentation3D, createStructureRepresentationParams(this.plugin, update.selector.data, {
            type: 'cartoon',
            color: 'uniform',
            colorParams: { value: color }
        }));
    
    //
    // Q: commit here or just close the updates by a single commit?
    //
    //await update.commit();
}

getQuery(chainId: string, auth_array: number[]): Expression {
    const query: Expression =
        MS.struct.generator.atomGroups({
            'residue-test': MS.core.set.has([MS.set( ...auth_array ), MS.ammp('auth_seq_id')]),
            'chain-test': MS.core.rel.eq([chainId, MS.ammp('label_asym_id')])
        });
    return query;
}
```

**Comment by dsehnal** [comment]
Hi,

maybe you can go back to your working version and just include the membrane representation on top of that?

```ts
if (!MembraneOrientationProvider.get(structure).value) {
    await plugin.runTask(Task.create('Membrane Orientation', async runtime => {
        await MembraneOrientationProvider.attach({ runtime, assetManager: plugin.managers.asset }, structure);
    }));
}

const membraneOrientation = await tryCreateMembraneOrientation(plugin, structureCell);
```

(taken from [here](https://github.com/molstar/molstar/blob/master/src/extensions/anvil/behavior.ts#L156))

**Comment by cycle20** [comment]
Hi,
Thank you very much. 
Sorry, I did not append some details:

- The membrane representation that you suggested earlier works very well. :+1: 
- My painting process was not perfect, because it added extra structures/items to the state tree.
  - Do I need to avoid those extra items in the tree?
  - Can they cause any performance degradation?
  - Or are they absolutely necessary parts of representation updates?
- I have two working representation updates, but I would like unite them in a single process with the least side(?) effect on the state tree. (NOTE: I do not know what is the difference between the states and structures.)

So I tried to eliminate it by sharing the structure (in Step 4 above):

```ts
   const structure: StateObjectRef<PluginStateObject.Molecule.Structure> =
            this.plugin.managers.structure.hierarchy.current.models[0].structures[0].cell;
```
Is it a right way to access that structure object? Or is there any specific other one tailored for this or similar cases?

> maybe you can go back to your working version and just include the membrane representation on top of that?

- Does it mean that the orders of the representation updates is important?
- Do colors, transparency, 3D representation of structures etc. have some kind of precedence?

> ```ts
> if (!MembraneOrientationProvider.get(structure).value) {
>     await plugin.runTask(Task.create('Membrane Orientation', async runtime => {
>         await MembraneOrientationProvider.attach({ runtime, assetManager: plugin.managers.asset }, structure);
>     }));
> }
> 
> const membraneOrientation = await tryCreateMembraneOrientation(plugin, structureCell);
> ```

- Do tasks ensure atomic updates on states and representations?
- Does this way provide some kind of queuing of updates and exclude parallel changes that cause unintended interference?
I will try this ```runTask```-```tryCreateMembraneOrientation```  as well.

**Comment by dsehnal** [comment]
With the coloring you have basically two options:

- Write a custom theme that colors the structure based on your data. For a custom theme you might want to check for example the [structure quality report extension](https://github.com/molstar/molstar/tree/master/src/extensions/pdbe/structure-quality-report)
- Use "overpaint". Here are some [helpers for it](https://github.com/molstar/molstar/blob/2dd07bb0e386358306660e89780de82a1b1f5c98/src/mol-plugin-state/helpers/structure-overpaint.ts).

> Does it mean that the orders of the representation updates is important?

No it isn't. What I meant you said you had a working version and then started adding the membranes, so I thought you would go back to it and add just the membrane representation "by hand".

> Do colors, transparency, 3D representation of structures etc. have some kind of precedence?

No.

> Do tasks ensure atomic updates on states and representations?

The ``await`` keyword does this. There queuing on the level of the tree updates, but doing concurrent changes can result in weird state, so it's best to do the queuing of state updates on the level of your code.

``pluing.runTask`` just uses the build in "task context" that displays the progress of the task in the default UI.

**Comment by cycle20** [comment]
Thank you very much for your quick support and detailed answers.
I'm closing this question and I will follow your theme guidelines.

NOTE:
Last night I could revert back to a previous working prototype.
I tried to repeat my refactoring changes step-by-step to catch the incorrect modification in the chain of updates with no luck.
I replaced my "handsome" abstractions (Region and descriptor interfaces) with a simple ```any``` keyword and I could proceed with the raw JSON from the backend. Something somehow interfered with the iterations and the parameter passing. The level of my TypeScript knowledge is almost zero and needs some improvement. :)
